File: random-walker/lua/random_walker.lua
#!/usr/bin/env lua
local json = require("json")

-- Unbuffered stdout so the runner sees each move immediately
io.stdout:setvbuf("no")
io.stderr:setvbuf("no")

-- Deterministic RNG
math.randomseed(1)

local first_tick = true
local moves = { "N", "S", "E", "W" }

for line in io.lines() do
  local ok, data = pcall(json.decode, line)
  if first_tick and type(data) == "table" and type(data.config) == "table" then
      local w = data.config.width or "?"
      local h = data.config.height or "?"
      io.stderr:write(string.format("Random walker (Lua) launching on a %sx%s map\n", w, h))
  end
  print(moves[math.random(1, #moves)])
  first_tick = false
end

File: random-walker/ruby/random-walker.rb
#!/usr/bin/env ruby
require 'json'

STDOUT.sync = STDERR.sync = true
srand(1)

first_tick = true

loop do
    line = STDIN.readline
    data = JSON.parse(line)
    if first_tick
        config = data['config']
        width = config['width']
        height = config['height']
        STDERR.puts "Random walker (Ruby) launching on a #{width}x#{height} map"
    end
    puts %w[N S E W].sample
    first_tick = false
end
